# Class 11: Derivatives 

Ch. 5, sections 10 - 12: Derivatives, Second derivatives, & Interpolation    

- Complete the activities as instructed by the professor

#### Import any packages we need below (update as we go):

In [1]:
import numpy as np
import math

### Forward, Backward, and Central difference 

Evaluate the derivative of $f(x) = e^{\sin(2x)}$ at $x = 0.5$

If you took the derivative by hand, what would it be? 

$$
\begin{align*}
f'(x) &= 2\cos(2x)e^{\sin(2x)}\\
f'(0.5) &= 2\cos(1)e^{\sin(1)}\\
&\approx2.504~\text{by hand}
\end{align*}
$$

In [5]:
def fprime(x):
    return 2*np.cos(2*x)*np.exp(np.sin(2*x))

true_val = fprime(0.5)
print(true_val)

2.506761534986894


Now apply forward/backward difference 

In [16]:
def f(x):
    return np.exp(np.sin(2*x))


#forward difference
def calc_fd(f,x,h):
    return (f(x+h)-f(x))/h


#define vars
x = 0.5
h_fd = [10**(-i) for i in range(1,12)]
#define an array of fd values
fd_array = np.zeros(len(h_fd))

for j in range(len(fd_array)):
    fd_array[j] = calc_fd(f,x,h_fd[j])

fd_error = np.abs(fd_array - true_val)



#backward difference
def calc_bd(f,x,h):
    return (f(x)-f(x-h))/h

h_bd = [10**(-i) for i in range(1,12)]
bd_array = np.zeros(len(h_bd))

for j in range(len(bd_array)):
    bd_array[j] = calc_bd(f,x,h_bd[j])

bd_error = np.abs(bd_array - true_val)





print("the true value is ", true_val)
print(" h      forward diff        error           backward diff        error")
rowf = "{0:1.0e} {1:1.16f} {2:1.16f} {3:1.16f} {4:1.16f}" 
for i in range(len(h_fd)):
    print(rowf.format(h_fd[i],fd_array[i],fd_error[i],bd_array[i],bd_error[i]))


the true value is  2.506761534986894
 h      forward diff        error           backward diff        error
1e-01 2.1990570766492690 0.3077044583376249 2.7076817455157842 0.2009202105288903
1e-02 2.4807256192968197 0.0260359156900742 2.5317171676925376 0.0249556327056437
1e-03 2.5042064928371133 0.0025550421497806 2.5093057730525281 0.0025442380656342
1e-04 2.5065065168927703 0.0002550180941236 2.5070164450369248 0.0002549100500309
1e-05 2.5067360380326420 0.0000254969542519 2.5067870308426166 0.0000254958557226
1e-06 2.5067589857208361 0.0000025492660578 2.5067640843090544 0.0000025493221605
1e-07 2.5067612785534266 0.0000002564334673 2.5067617892560179 0.0000002542691240
1e-08 2.5067615538887367 0.0000000189018428 2.5067615538887367 0.0000000189018428
1e-09 2.5067619091601046 0.0000003741732106 2.5067614650708947 0.0000000699159992
1e-10 2.5067636855169440 0.0000021505300500 2.5067592446248455 0.0000022903620485
1e-11 2.5067947717616335 0.0000332367747395 2.5067503628406484 0.0000111

Now apply central difference 

In [17]:
#central difference
def calc_cd(f,x,h):
    return (f(x+(h/2))-f(x-(h/2)))/h


#define vars
x = 0.5
h_cd = [10**(-i) for i in range(1,12)]
#define an array of cd values
cd_array = np.zeros(len(h_cd))

for j in range(len(cd_array)):
    cd_array[j] = calc_cd(f,x,h_cd[j])

cd_error = np.abs(cd_array - true_val)

print("the true value is ", true_val)
print(" h      central diff        error")
rowf = "{0:1.0e} {1:1.16f} {2:1.16f}"
for i in range(len(h_cd)):
    print(rowf.format(h_cd[i],cd_array[i],cd_error[i]))


the true value is  2.506761534986894
 h      central diff        error
1e-01 2.4932959255171205 0.0134656094697734
1e-02 2.5066264877375843 0.0001350472493096
1e-03 2.5067601844748211 0.0000013505120728
1e-04 2.5067615214791061 0.0000000135077878
1e-05 2.5067615348373096 0.0000000001495843
1e-06 2.5067615352369899 0.0000000002500959
1e-07 2.5067615361251683 0.0000000011382744
1e-08 2.5067615538887367 0.0000000189018428
1e-09 2.5067614650708947 0.0000000699159992
1e-10 2.5067636855169440 0.0000021505300500
1e-11 2.5067503628406484 0.0000111721462455


How do they compare? 

### Second derivatives

Let's take the second derivative use central difference of the same function $f(x) = e^{\sin(2x)}$ at $x = 0.5$ & compare with the true value 

$$ \frac{d^2}{dx^2}e^{\sin(2x)} = 2 e^{\sin(2x)}(-2\sin(2x)+\cos(4x)+1)  $$

In [19]:
def f(x):
    return np.exp(np.sin(2*x))

def d2f(x):
    return 2*np.exp(np.sin(2*x))*(-2*np.sin(2*x)+np.cos(4*x)+1)

true_val_doub = d2f(0.5)


#define constants
x = 0.5
h = 1e-4 #ideal h for second derivative

def calc_cd(f,x,h):
    return (f(x+(h/2))-f(x-(h/2)))/h


d1xplush2 = calc_cd(f,x+(h/2),h)

d1xminush2 = calc_cd(f,x-(h/2),h)

d2x = (d1xplush2 - d1xminush2) / h

d2error = d2x - true_val_doub



print("the true value is ", true_val_doub)
print(" h      second diff        error")
rowf = "{0:1.0e} {1:1.16f} {2:1.16f}"
print(h,d2x,d2error)


the true value is  -5.099281481682785
 h      second diff        error
0.0001 -5.099281441545145 4.013763987842367e-08


### Automatic Differentiation - book example

Store value/derivative pairs as two-element lists in Python of the form [value, derivative]

Define some functions to perform basic mathematics operations on these inputted lists. 

Apply this to overall figure out the value of the function $f(x) = 3x^2 +e^x$ and its derivative at $t = 1/2$

To do this, perform the following sequence of operations: 

$$ u_1 = x^2$$
$$ u_2 = 3u_1$$
$$ u_3 = e^x$$

and finally 

$$ f(x) = u_2 + u_3$$

Each each mathematical operation will be performed using the generalized operations listed above and the end result will be a pair $[f,f']$, representing the value of $f$ and its derivative.

Final bit of info: to calculate the derivative $f'(t)$ at any value of $t$, need an input list to evaluate our function $f(𝑥)$. Use $[t,1]$ since for whatever value of $t$ we are interested in, by definition, this is the correct value/derivative pair for the independent variable $t$, because $dt/dt = 1$ always. So if we use this pair as the argument of our function, we will get out a pair representing the value of the function $f(t)$ and its derivative $f'(t)$, at the chosen value of $t$.


In [24]:
def power(x,c):
    vx,dx = x
    return [vx**c, c*vx**(c-1)*dx]

def mult(x,c):
    vx,dx = x
    return [c*vx,c*dx]

def exp(x):
    vx,dx = x
    return [math.exp(vx),dx*math.exp(vx)]

def add(x,y):
    vx,dx = x
    vy,dy = y
    return [vx+vy,dx+dy]


def autodiff(x):
    u1 = power(x,2)
    u2 = mult(u1,3)
    u3 = exp(x)
    u4 = add(u2,u3)
    return u4

print(autodiff([0.5,1]))













[2.398721270700128, 4.648721270700128]


The actual value of the function and the derivative at $f(t)$ at $t = 1/2$ is 
$$f (t) = 3t^2 + e^t = \frac{3}{4} + e^{1/2} = 2.3987 ...$$
$$ f'(t) = 6t + e^t = 3 + e^{1/2} = 4.6487 ...$$

Look how exact we are!